### PR TITLE
refactor: refactor hash join spill

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

We should reset the number of kinds of workers to the number of active processors at once, rather than adding one per processor.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12894)

### CODE REVIEW

**Comment by BohuTANG** [comment]
Can we bind the `groupby_spilled_to_storage_bytes` or `join_spilled_to_storage_bytes` to the query session progress? So we can store them to the system.query_log.

**Comment by xudong963** [comment]
> Can we bind the `groupby_spilled_to_storage_bytes` or `join_spilled_to_storage_bytes` to the query session progress? So we can store them to the system.query_log.

Yeah, let's do it in next PR.

**Review by zhang2014 — APPROVED** [review]

File: src/query/service/src/pipelines/processors/transforms/hash_join/transform_hash_join_build.rs [hunk 1]
```diff
@@ -102,13 +104,14 @@ impl TransformHashJoinBuild {
...
     // It means next round build will start, need to reset some variables.
     fn reset(&mut self) -> Result<()> {
-        let mut count = self.build_state.row_space_builders.lock();
-        *count += 1;
-        let mut count = self.build_state.hash_join_state.hash_table_builders.lock();
         self.finalize_finished = false;
-        *count += 1;
         // Only need to reset the following variables once
-        if *count == 1 {
+        if *self.build_state.row_space_builders.lock() == 0 {
```
Comment (Dousir9) [inline src/query/service/src/pipelines/processors/transforms/hash_join/transform_hash_join_build.rs:109]: There may be two threads enter this statement block at the same time and get the same build_worker_num.

**Comment by xudong963** [comment]
```
failures:
3152

3153
---- grpc::metasrv_grpc_watch::test_watch_expired_events stdout ----
3154
thread 'grpc::metasrv_grpc_watch::test_watch_expired_events' panicked at 'assertion failed: `(left == right)`
3155
  left: `Some(Event { key: "w_auto_gc_0", current: None, prev: Some(SeqV { seq: 1, data: [119, 95, 97, 117, 116, 111, 95, 103, 99, 95, 48] }) })`,
3156
 right: `Some(Event { key: "w_b1", current: None, prev: Some(SeqV { seq: 34, data: [119, 95, 98, 49] }) })`', src/meta/service/tests/it/grpc/metasrv_grpc_watch.rs:446:13
3157
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
https://github.com/datafuselabs/databend/actions/runs/6258675682/job/16993150945?pr=12894 cc @drmingdrmer

**Comment by drmingdrmer** [comment]
> ```
> failures:
> 3152
> 
> 3153
> ---- grpc::metasrv_grpc_watch::test_watch_expired_events stdout ----
> 3154
> thread 'grpc::metasrv_grpc_watch::test_watch_expired_events' panicked at 'assertion failed: `(left == right)`
> 3155
>   left: `Some(Event { key: "w_auto_gc_0", current: None, prev: Some(SeqV { seq: 1, data: [119, 95, 97, 117, 116, 111, 95, 103, 99, 95, 48] }) })`,
> 3156
>  right: `Some(Event { key: "w_b1", current: None, prev: Some(SeqV { seq: 34, data: [119, 95, 98, 49] }) })`', src/meta/service/tests/it/grpc/metasrv_grpc_watch.rs:446:13
> 3157
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> https://github.com/datafuselabs/databend/actions/runs/6258675682/job/16993150945?pr=12894 cc @drmingdrmer

It's a timeout that caused watching event loss. Just ignore it.

**Comment by BohuTANG** [comment]
```
[SQL] select a from t3 inner join numbers(100000) on t3.a = number order by a;
[Diff] (-expected|+actual)
-   0
    1
-   2
-   3
-   4
-   5
    6
-   7
-   8
-   9
+   8
at tests/sqllogictests/suites/query/spill.test:27
```

**Comment by xudong963** [comment]
> ```
> [SQL] select a from t3 inner join numbers(100000) on t3.a = number order by a;
> [Diff] (-expected|+actual)
> -   0
>     1
> -   2
> -   3
> -   4
> -   5
>     6
> -   7
> -   8
> -   9
> +   8
> at tests/sqllogictests/suites/query/spill.test:27
> ```

The bug is fixed and the test file has been run locally hundreds of times. : )

**Pull request merged**